contrib/pyzfs/libzfs_core/test/test_libzfs_core.py

```diff
@@ -0,0 +1,3731 @@
+            bad_fd = tmp.fileno()
+
+        (soft, hard) = resource.getrlimit(resource.RLIMIT_NOFILE)
+        bad_fd = hard + 1
+        with self.assertRaises(lzc_exc.StreamIOError) as ctx:
+            lzc.lzc_send(snap, None, bad_fd)
+        self.assertEquals(ctx.exception.errno, errno.EBADF)
+
+    def test_send_to_broken_pipe(self):
+        snap = ZFSTest.pool.makeName("fs1@snap")
+        lzc.lzc_snapshot([snap])
+
+        proc = subprocess.Popen(['true'], stdin=subprocess.PIPE)
+        proc.wait()
+        with self.assertRaises(lzc_exc.StreamIOError) as ctx:
+            lzc.lzc_send(snap, None, proc.stdin.fileno())
+        self.assertEquals(ctx.exception.errno, errno.EPIPE)
+
+    def test_send_to_broken_pipe_2(self):
+        snap = ZFSTest.pool.makeName("fs1@snap")
+        with zfs_mount(ZFSTest.pool.makeName("fs1")) as mntdir:
+            with tempfile.NamedTemporaryFile(dir=mntdir) as f:
+                for i in range(1024):
+                    f.write('x' * 1024)
+                f.flush()
+                lzc.lzc_snapshot([snap])
+
+        proc = subprocess.Popen(['sleep', '2'], stdin=subprocess.PIPE)
+        with self.assertRaises(lzc_exc.StreamIOError) as ctx:
+            lzc.lzc_send(snap, None, proc.stdin.fileno())
+        self.assertTrue(ctx.exception.errno == errno.EPIPE or
+                        ctx.exception.errno == errno.EINTR)
+
+    def test_send_to_ro_file(self):
+        snap = ZFSTest.pool.makeName("fs1@snap")
+        lzc.lzc_snapshot([snap])
+
+        with tempfile.NamedTemporaryFile(
+                suffix='.ztream', delete=False) as output:
+            # tempfile always opens a temporary file in read-write mode
+            # regardless of the specified mode, so we have to open it again.
+            os.chmod(output.name, stat.S_IRUSR)
+            fd = os.open(output.name, os.O_RDONLY)
+            with self.assertRaises(lzc_exc.StreamIOError) as ctx:
+                lzc.lzc_send(snap, None, fd)
+            os.close(fd)
+        self.assertEquals(ctx.exception.errno, errno.EBADF)
+
+    def test_recv_full(self):
+        src = ZFSTest.pool.makeName("fs1@snap")
+        dst = ZFSTest.pool.makeName("fs2/received-1@snap")
+
+        with temp_file_in_fs(ZFSTest.pool.makeName("fs1")) as name:
+            lzc.lzc_snapshot([src])
+
+        with tempfile.TemporaryFile(suffix='.ztream') as stream:
+            lzc.lzc_send(src, None, stream.fileno())
+            stream.seek(0)
+            lzc.lzc_receive(dst, stream.fileno())
+
+        name = os.path.basename(name)
+        with zfs_mount(src) as mnt1, zfs_mount(dst) as mnt2:
+            self.assertTrue(
+                filecmp.cmp(
+                    os.path.join(mnt1, name), os.path.join(mnt2, name), False))
+
+    def test_recv_incremental(self):
+        src1 = ZFSTest.pool.makeName("fs1@snap1")
+        src2 = ZFSTest.pool.makeName("fs1@snap2")
+        dst1 = ZFSTest.pool.makeName("fs2/received-2@snap1")
```

Comment:
This test is failing on current master but passes on zfs-0.6.5.11:
 
```FAIL: test_hold_missing_fs (libzfs_core.test.test_libzfs_core.ZFSTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/libzfs_core/test/test_libzfs_core.py", line 431, in test_hold_missing_fs
    lzc.lzc_hold({snap: 'tag'})
AssertionError: HoldFailure not raised
```

Actually this seems to be caused by "_OpenZFS 7071 - lzc_snapshot does not fill in errlist on ENOENT_" (e88551d52f0490578ec6d2aa2c36563dfdbdd3b3): before this commit `lzc_hold()` will return ENOENT which raises the expected `HoldFailure` in `test_hold_missing_fs()`, but the ENOENT is actually from `zfs_secpolicy_write_perms()`, not `zfs_ioc_hold()`.

After said commit both `zfs_secpolicy_write_perms()` and `zfs_ioc_hold()` return 0, and so does `lzc_hold()`: `HoldFailure` doesn't get raised in the Python code and thus the test fails:

```
root@linux:~# cat /sys/module/zfs/version 
0.7.0-339_g2a0428f
root@linux:~# cat > ./zfs_hold.sh <<'EOF'
> POOLNAME='testpool'
> TMPDIR='/var/tmp'
> mountpoint -q $TMPDIR || mount -t tmpfs tmpfs $TMPDIR
> zpool destroy -f $POOLNAME
> rm -f $TMPDIR/zpool.dat
> fallocate -l 128m $TMPDIR/zpool.dat
> zpool create -f $POOLNAME $TMPDIR/zpool.dat
> zfs create $POOLNAME/fs1
> zfs snap $POOLNAME/fs1@snap1
> python -c "import libzfs_core as lzc; print lzc.lzc_hold({'$POOLNAME/fs2@snap2': 'tag'})"
> EOF
root@linux:~# chmod +x ./zfs_hold.sh
root@linux:~# 
root@linux:~# stap -d zfs -e '
> probe
> module("zfs").function("zfs_ioc_hold").return,
> module("zfs").function("zfs_secpolicy_write_perms").return,
> module("zfs").function("dsl_dataset_user_hold").return,
> module("zfs").function("dsl_dataset_hold").return
> { printf(" <- %s ret=%s\n", ppfunc(), $$return$$); }' -c ./zfs_hold.sh
 <- zfs_secpolicy_write_perms ret=return=0
 <- zfs_secpolicy_write_perms ret=return=0
 <- zfs_secpolicy_write_perms ret=return=0
['testpool/fs2@snap2']
 <- zfs_secpolicy_write_perms ret=return=0
 <- dsl_dataset_hold ret=return=2
 <- dsl_dataset_user_hold ret=return=0
 <- zfs_ioc_hold ret=return=0
root@linux:~# 
```